### ISSUE TITLE
Calculation Groups with Columns in "Ordinal, Name" order return confusing data type comparison error

### ISSUE THREAD

**Issue by o-o00o-o** [otykier/TabularEditor#411]
We have stumbled upon another really weird bug. I have build a repro that I've attached but it needs some explanation.

A new Calculation Group (our 3rd one) was resulting in an error "DAX comparison operations do not support comparing values of type Integer with values of type Text" when we were trying to use it. The first 2 still worked fine.

It seems that the Calculation Group table itself is "contaminated" in a way that seems invisible in the metadata and no matter what you do it will not work.

Not sure at this point whether this is a Tabular Editor issue or AAS

I built a repro that is a simple standalone datasource from some static power query data, has one table and 3 calc groups. CG1 and CG2 are "good", "Target Calculations" is "bad". I couldn't create the issue using the normal tabular editor tools, so I copied "Target Calculations" from our real model by copying the json files and interestingly it is still "bad".

steps to repro
- Open my repro bim attached [TroubleshootCalcGroupsIssue.zip](https://github.com/otykier/TabularEditor/files/4395421/TroubleshootCalcGroupsIssue.zip)
- deploy it to AAS (our v shows as 15.1.27.20) as a new Database using Tabular Editor 2.9.5
- Process the database (it is using static data in the model so it should just work)

``` dax
{
  "refresh": {
    "type": "full",
    "objects": [
      {
        "database": "BrettTest"
      }
    ]
  }
}
```
- Now run the following queries

```

--# this works
evaluate row("blah", sum(Orders[Price]))

--# this works!
EVALUATE
SUMMARIZECOLUMNS(
"Blah", CALCULATE(
    [OrderPrice],
    'CG1'[CG1] = "Actual"
    )
)

--# this works!
EVALUATE
SUMMARIZECOLUMNS(
"Blah", CALCULATE(
    [OrderPrice],
    'CG2'[CG2] = "Actual"
    )
)

--# doesn't work! returns "DAX comparison operations do not support comparing values of type Integer with values of type Text" 
EVALUATE
SUMMARIZECOLUMNS(
"Blah", CALCULATE(
    [OrderPrice],
    'Target Calculations'[Target Calculations] = "Actual"
    )
)

```

Interestingly 
 - If you now connect tabular editor to the database and copy a good CG, this will work, if you copy the bad CG (Target Calculations) it will not work. It's like the table is contaminated although diffing shows no significant differences in the metadata (easier when serialised)
- Previously deployed CGs (using Tabular Editor) worked fine, this 3rd one has caused us this problem
- If we connect to the DB and add the Calculation Group there and Save it, it works fine, we just can't deploy it from Code
- We are running on AAS. Cube in Folder serialisation, but as you can see the bim also repro's the problem
- I've spent about 3 hours trying to repro this and trying different things so I've run out of time. I guess another thing to try would be to import the BIM into VS and see if it is a AAS or a Tabular Editor deploy issue.

**Comment by o-o00o-o** [comment]
So, I couldn't leave this one alone.. too weird, and I finally got to the bottom of it.

By default, Calculation Groups (CG) have two columns, Name and Ordinal. Name is a bad name for the CG Column as most tools (e.g. Power BI) will only show the column name. So if a user have more than one CG applying, they will not be able to easily see what it is (they will all be "Name"). So our convention is to rename the Name column to be the same as the CG name. 

The problem is that, AAS is expecting to be presented the Columns in the order Name then Ordinal in the script. But when I rename it to something that makes it alphabetically after "Ordinal", it reverses the order and creates this error.

Our previous two CG's were "Date Calculations" and "Like for Like Calculations", so as these were alphabetically before "Ordinal" these didn't hit this issue which is why it stumped me.

If I manually change the script to put the columns so that the "Name" column is before the "Ordinal" column, it fixes the issue

![image](https://user-images.githubusercontent.com/13693159/77811177-a7bcef00-7090-11ea-82e3-1e81db6f1240.png)

For now I have renamed the Ordinal column to be zOrdinal to ensure it is always after. It seems to work for now but it feels a bit kludgy.

Perhaps the script generator in Tabular Editor could change the order of CG columns to order them by "Source Column" order rather than "Name" order? As they have to have Name and Ordinal, this will ensure they are always in the correct order. 

Now maybe I can get to sleep! :)

**Comment by otykier** [comment]
Whoa, great find!

I’ll report this as a bug with the product team (as the metadata order of columns should really be irrelevant). But since it’s a very minor fix on my side, I’ll also make sure that Tabular Editor serializes the Name column before the Ordinal column for calc groups. Expect a new release this Monday.

**Comment by o-o00o-o** [comment]
Awesome, thanks Daniel

Brett

**Comment by otykier** [comment]
Fixed in [2.9.6](https://github.com/otykier/TabularEditor/releases/tag/2.9.6)

**Comment by helgews** [comment]
Managed to get this bug again today, deploying from Tabular Editor 2.14.2.
I took a duplicate from one of my existing calculation groups, renamed the column to something alphabetically after "Ordinal" and deployed it.

The solution was to delete the calculation group, deploy, undo deletion, choose column name which is alphabetically before "Ordinal", deploy again.

Thank you for the "zOrdinal"-tip, by the way! Now I can use my whole vocabulary for my column-names again :)
